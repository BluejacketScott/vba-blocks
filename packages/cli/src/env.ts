import { join } from 'path';
import { homedir } from 'os';
import { Reporter, reporter } from './reporter';

export interface Env {
  isWindows: boolean;
  cwd: string;
  values: any;

  addins: string;
  scripts: string;
  cache: string;
  registry: string;
  packages: string;
  sources: string;
  staging: string | null;

  reporter: Reporter;
}

const cache = join(homedir(), '.vba-blocks');
const env: Env = {
  isWindows: process.platform === 'win32',
  cwd: process.cwd(),
  values: process.env,

  addins: join(__dirname, '../../addin/build'),
  scripts: join(__dirname, '../scripts'),
  cache,
  registry: join(cache, 'registry'),
  packages: join(cache, 'packages'),
  sources: join(cache, 'sources'),
  staging: null,

  reporter
};

export default env;
