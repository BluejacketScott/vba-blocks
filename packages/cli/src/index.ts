export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
