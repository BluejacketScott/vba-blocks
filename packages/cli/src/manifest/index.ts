import { join } from 'path';
import { ok } from 'assert';
import { parse as parseToml } from 'toml';
import { pathExists, readFile } from '../utils/fs';
import { Version } from './version';
import { Source, parseSrc } from './source';
import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';

export { Version, Source, Feature, Dependency, Reference, Target };

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 * 
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 * 
 * [features]
 * default = ["a"]
 * 
 * a = { src = ["C"] }
 * b = { dependencies = ["dictionary"] }
 * c = { references = ["Scripting"] }
 * 
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-features = { version = "1.0.0", default-features = false, features = ["other"] }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 * 
 * [references.Scripting]
 * version = "1.0"
 * guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 * optional = true
 * 
 * [[targets]]
 * type = "xlsm"
 * path = "targets/xlsm"
 * 
 * [[targets]]
 * name = "custom-name"
 * type = "xlam"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  features: Feature[];
  defaultFeatures: string[];
  dir: string;
}

const EXAMPLE = `Example vba-block.toml:

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  ok(
    value && value.package,
    `[package] is a required field, with name, version, and authors specified. ${EXAMPLE}`
  );

  const { name, version, authors, publish = false } = value.package;

  ok(name, `[package] name is a required field. ${EXAMPLE}`);
  ok(version, `[package] version is a required field. ${EXAMPLE}`);
  ok(authors, `[package] authors is a required field. ${EXAMPLE}`);

  const src = parseSrc(value.src || {}, dir);
  const { features, defaultFeatures } = parseFeatures(value.features || {});
  const dependencies = parseDependencies(value.dependencies || {});
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || [], name, dir);

  const metadata = { publish, ...value.package };

  return {
    name,
    version,
    metadata,
    src,
    features,
    defaultFeatures,
    dependencies,
    references,
    targets,
    dir
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');
  if (!await pathExists(file)) {
    throw new Error(`vba-blocks.toml not found in "${dir}"`);
  }

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());
  const manifest = parseManifest(parsed, dir);

  return manifest;
}
