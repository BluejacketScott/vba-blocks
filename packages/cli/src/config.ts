import { join } from 'path';
import { parse as parseToml } from 'toml';
import env from './env';
import { pathExists, readFile } from './utils';
import {
  Sources,
  RegistrySource,
  PathSource,
  GitSource,
  UnsupportedSource
} from './sources';

export type Registry =
  | {}
  | { [name: string]: { index: string; packages: string } };

export interface Flags {
  git?: boolean;
  path?: boolean;
}

export interface Config {
  registry: Registry;
  flags: Flags;
  sources: Sources;
}

export interface ConfigValue {
  registry?: Registry;
  flags?: Flags;
}

const empty: ConfigValue = { registry: {}, flags: {} };
const defaults: ConfigValue = {
  registry: {
    'vba-blocks': {
      index: 'https://github.com/vba-blocks/registry',
      packages: 'https://packages.vba-blocks.com'
    }
  },
  flags: { git: true, path: true }
};

/**
 * Load config, from local, user, and environment values
 *
 * - Search for .vba-blocks/config.toml up from cwd
 * - Load ~/.vba-blocks/config.toml
 * - Load VBA_BLOCKS_* from environment
 */
export async function loadConfig(): Promise<Config> {
  const user: ConfigValue = {
    ...empty,
    ...((await readConfig(env.cache)) || {})
  };
  const file = await findConfig(env.cwd);
  const local: ConfigValue = {
    ...empty,
    ...file ? await readConfig(file) : {}
  };
  const override = loadConfigFromEnv();

  const registry: Registry = {
    ...defaults.registry,
    ...user.registry,
    ...local.registry,
    ...override.registry
  };
  const flags: Flags = {
    ...defaults.flags,
    ...user.flags,
    ...local.flags,
    ...override.flags
  };

  const sources: Sources = {
    registry: {},
    git: flags.git ? new GitSource() : new UnsupportedSource('git'),
    path: flags.path ? new PathSource() : new UnsupportedSource('path')
  };

  for (const [name, { index, packages }] of Object.entries(registry)) {
    sources.registry[name] = new RegistrySource({
      name,
      index,
      packages
    });
  }

  return { registry, flags, sources };
}

// Read config from dir (if present)
export async function readConfig(
  dir: string
): Promise<ConfigValue | undefined> {
  const file = join(dir, 'config.toml');
  if (!await pathExists(file)) return {};

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());

  return parsed;
}

// Find config up from and including given dir
// (looking for .vba-blocks/config.toml)
export async function findConfig(dir: string): Promise<string | undefined> {
  // TODO Search from .vba-blocks/config.toml starting at cwd
  return;
}

// Load config values from environment
export function loadConfigFromEnv(): ConfigValue {
  // TODO Load override config from VBA_BLOCKS_* env variables
  return {};
}
