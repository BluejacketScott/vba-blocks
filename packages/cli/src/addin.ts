import env from './env';
import { Project } from './project';
import { Target, Source, Reference } from './manifest';
import { run, unixJoin } from './utils';

export type Application = string;
export type Addin = string;

export const extensions: { [application: string]: string[] } = {
  excel: ['xlsx', 'xlsm', 'xlam']
};
export const addins: { [application: string]: string } = {
  excel: 'vba-blocks.xlsm' // TEMP vba-blocks.xlam
};

const byExtension: { [extension: string]: string } = {};
for (const [application, values] of Object.entries(extensions)) {
  for (const extension of values) {
    byExtension[extension] = application;
  }
}

/**
 * Import graph of src and references into given target
 */
export async function importGraph(
  project: Project,
  target: Target,
  graph: { src: Source[]; references: Reference[] }
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);
  const { src, references } = graph;

  await run(application, addin, 'Build.ImportGraph', {
    file,
    src,
    references
  });
}

/**
 * Export src and references from given target
 */
export async function exportTo(
  project: Project,
  target: Target,
  staging: string
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);

  const result = await run(application, addin, 'Build.ExportTo', {
    file,
    staging
  });
  console.log('export result', result);
}

/**
 * Get application, addin, and file for given target
 */
export function getTargetInfo(
  project: Project,
  target: Target
): { application: Application; addin: Addin; file: string } {
  const application = byExtension[target.type];
  if (!application) throw new Error(`Unsupported target type "${target.type}"`);

  const addin = unixJoin(env.addins, addins[application]);
  const file = unixJoin(project.paths.build, target.filename);

  return { application, addin, file };
}
