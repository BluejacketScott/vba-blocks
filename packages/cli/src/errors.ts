import chalk from 'chalk';
import env from './env';
import { cleanError, isString } from './utils';

import { Target } from './manifest';
import { Registration } from './sources';

export interface CliErrors {
  'unknown-command': (command: string) => string;

  'unsupported-source': (type: string) => string;
  'dependency-not-found': (dependency: string, registry: string) => string;
  'dependency-invalid-checksum': (registration: Registration) => string;

  'target-not-found': (target: Target) => string;
  'target-is-open': (target: Target, path: string) => string;
  'target-import-failed': (target: Target) => string;
  'target-restore-failed': (backup: string, file: string) => string;

  'resolve-failed': (details?: string) => string;

  'run-script-not-found': (path: string) => string;
}

export interface CliErrorOptions {
  code?: string;
  underlying?: Error;
}

export class CliError extends Error {
  code?: string;
  underlying?: Error;

  constructor(message: string, options: CliErrorOptions = {}) {
    if (options.underlying) {
      const underlying = cleanError(options.underlying);
      message += chalk`\n\n{dim ${underlying.message}}`;
    }

    super(message);

    this.code = options.code;
    this.underlying = options.underlying;
  }
}

// TODO Resolve typing issues so the following isn't so repetitive/verbose

export function unknownCommand(command: string): Error {
  const code = 'unknown-command';
  const message = env.reporter.errors[code](command);

  return new CliError(message, { code });
}

export function unsupportedSource(type: string): Error {
  const code = 'unsupported-source';
  const message = env.reporter.errors[code](type);

  return new CliError(message, { code });
}

export function dependencyNotFound(
  dependency: string,
  registry: string
): Error {
  const code = 'dependency-not-found';
  const message = env.reporter.errors[code](dependency, registry);

  return new CliError(message, { code });
}

export function dependencyInvalidChecksum(registration: Registration): Error {
  const code = 'dependency-invalid-checksum';
  const message = env.reporter.errors[code](registration);

  return new CliError(message, { code });
}

export function targetNotFound(target: Target): Error {
  const code = 'target-not-found';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code });
}

export function targetIsOpen(target: Target, path: string): Error {
  const code = 'target-is-open';
  const message = env.reporter.errors[code](target, path);

  return new CliError(message, { code });
}

export function targetImportFailed(target: Target, underlying: Error): Error {
  const code = 'target-import-failed';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code, underlying });
}

export function targetRestoreFailed(
  backup: string,
  file: string,
  underlying: Error
): Error {
  const code = 'target-restore-failed';
  const message = env.reporter.errors[code](backup, file);

  return new CliError(message, { code, underlying });
}

export function resolveFailed(details?: string): Error {
  const code = 'resolve-failed';
  let message = env.reporter.errors[code]();
  if (details) {
    message += `\n${details}`;
  }

  return new CliError(message, { code });
}

export function runScriptNotFound(path: string): Error {
  const code = 'run-script-not-found';
  const message = env.reporter.errors[code](path);

  return new CliError(message, { code });
}
