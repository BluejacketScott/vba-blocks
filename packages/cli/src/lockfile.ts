import { ok } from 'assert';
import { join, relative } from 'path';
import { parse as parseToml } from 'toml';
import { has, convertToToml, unixPath } from './utils';
import { pathExists, readFile, writeFile } from './utils/fs';
import { Snapshot, Manifest } from './manifest';
import { Dependency, satisfies } from './manifest/dependency';
import { Workspace } from './workspace';
import {
  Registration,
  getRegistrationId,
  getRegistrationSource,
  getSourceParts,
  toDependency
} from './sources/registration';
import { DependencyGraph, getRegistration } from './resolve';

export interface Lockfile {
  workspace: {
    root: Snapshot;
    members: Snapshot[];
  };
  packages: DependencyGraph;
}

type DependencyByName = Map<string, Dependency>;

/**
 * Read lockfile at given dir (if present)
 * (for invalid lockfile, errors are ignored and treated as no lockfile)
 */
export async function readLockfile(dir: string): Promise<Lockfile | null> {
  try {
    const file = join(dir, 'vba-block.lock');
    if (!await pathExists(file)) return null;

    const toml = await readFile(file, 'utf8');
    return fromToml(toml, dir);
  } catch (err) {
    // TODO Log error

    return null;
  }
}

/**
 * Write lockfile for project to given dir
 */
export async function writeLockfile(
  dir: string,
  lockfile: Lockfile
): Promise<void> {
  const file = join(dir, 'vba-block.lock');
  const toml = toToml(lockfile, dir);

  return writeFile(file, toml);
}

// Check if lockfile is still valid for loaded workspace
// (e.g. invalidated by changing/adding dependency to manifest)
export function isLockfileValid(
  lockfile: Lockfile,
  workspace: Workspace
): boolean {
  if (!compareDependencies(workspace.root, lockfile.workspace.root))
    return false;

  if (lockfile.workspace.members.length !== workspace.members.length)
    return false;

  const byName: { [name: string]: Snapshot } = {};
  workspace.members.forEach(member => (byName[member.name] = member));

  for (const member of lockfile.workspace.members) {
    const currentMember = byName[member.name];
    if (!currentMember) return false;
    if (!compareDependencies(currentMember, member)) return false;
  }

  return true;
}

// Convert lockfile/project to toml
// - toml, alphabetized and with trailing commas, should be suitable for VCS
export function toToml(lockfile: Lockfile, dir: string): string {
  const root = prepareManifest(lockfile.workspace.root, lockfile.packages, dir);
  const members: any[] = lockfile.workspace.members.map((member: Snapshot) =>
    prepareManifest(member, lockfile.packages, dir)
  );
  const packages: any[] = lockfile.packages.map(
    (registration: Registration) => {
      const { name, version, source } = registration;
      const dependencies = registration.dependencies.map(dependency =>
        toDependencyId(dependency, lockfile.packages, dir)
      );

      return {
        name,
        version,
        source: prepareSource(source, dir),
        dependencies
      };
    }
  );

  return convertToToml({ root, members, packages });
}

// Load lockfile from toml (including "hydrating" dependencies from packages)
export function fromToml(toml: string, dir: string): Lockfile {
  const parsed = parseToml(toml);
  ok(has(parsed, 'root'), 'vba-block.lock is missing [root] field');

  // First pass through packages to load high-level information
  // (needed to map dependencies to parsed packages)
  const byName: DependencyByName = new Map();
  const packages = (parsed.packages || []).map((value: any) => {
    const { name, version, source, dependencies } = value;
    ok(
      name && version && source && Array.isArray(dependencies),
      'Invalid package in lockfile'
    );

    const registration = {
      id: getRegistrationId(name, version),
      name,
      version,
      source: getSource(source, dir),
      dependencies
    };

    byName.set(name, toDependency(registration));

    return registration;
  });

  // Hydrate dependencies of packages
  packages.forEach((registration: any) => {
    registration.dependencies = registration.dependencies.map((id: string) =>
      getDependency(id, byName)
    );
  });

  // Load manifests for workspace
  const root = toManifest(parsed.root, byName);
  const members = (parsed.members || []).map((member: any) =>
    toManifest(member, byName)
  );

  return { workspace: { root, members }, packages };
}

// Convert raw toml value to manifest
function toManifest(value: any, byName: DependencyByName): Snapshot {
  const { name, version } = value;
  ok(
    name && version && Array.isArray(value.dependencies),
    'Invalid manifest in lockfile'
  );

  const dependencies = value.dependencies.map((id: string) =>
    getDependency(id, byName)
  );

  return {
    name,
    version,
    dependencies
  };
}

// Prepare manifest for toml
function prepareManifest(
  manifest: Snapshot,
  packages: DependencyGraph,
  dir: string
): any {
  const { name, version } = manifest;
  const dependencies = manifest.dependencies.map(dependency =>
    toDependencyId(dependency, packages, dir)
  );

  return {
    name,
    version,
    dependencies
  };
}

// Prepare registration source for toml
// (Convert full path to relative to root dir)
function prepareSource(source: string, dir: string): string {
  const { type, value, details } = getSourceParts(source);
  if (type !== 'path') return source;

  const relativePath = unixPath(relative(dir, value));
  return getRegistrationSource(type, relativePath, details);
}

// "Hydrate" registration source from toml
// (Convert relative paths to absolute)
function getSource(source: string, dir: string): string {
  const { type, value, details } = getSourceParts(source);
  if (type !== 'path') return source;

  const absolutePath = join(dir, value);
  return getRegistrationSource(type, absolutePath, details);
}

// Get dependency id
//
// Minimum information needed for lockfile:
// "{name} {version} {source}"
function toDependencyId(
  dependency: Dependency,
  packages: DependencyGraph,
  dir: string
) {
  const registration = getRegistration(packages, dependency);
  ok(registration, 'No package found for dependency');

  let { version, source } = registration!;
  const { type, value, details } = getSourceParts(source);

  if (type === 'path') {
    const relativePath = unixPath(relative(dir, value));
    source = getRegistrationSource(type, relativePath, details);
  }

  return `${dependency.name} ${version} ${source}`;
}

// Get dependency by id (using name from id)
function getDependency(id: string, byName: DependencyByName): Dependency {
  const [name] = id.split(' ', 1);
  const dependency = byName.get(name);

  ok(dependency, `Package not found in lockfile, "${id}"`);

  return dependency!;
}

// Compare dependencies between current user manifest and lockfile manifest
function compareDependencies(current: Snapshot, locked: Snapshot): boolean {
  if (current.dependencies.length !== locked.dependencies.length) return false;

  const byName: { [name: string]: Dependency } = {};
  current.dependencies.forEach(
    dependency => (byName[dependency.name] = dependency)
  );

  for (const dependency of locked.dependencies) {
    const currentValue = byName[dependency.name];
    if (!currentValue) return false;
    if (!satisfies(currentValue, dependency)) return false;
  }

  return true;
}
