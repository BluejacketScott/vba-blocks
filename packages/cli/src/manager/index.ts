import { Config } from '../config';
import { Dependency } from '../manifest';
import { isRegistryDependency, isGitDependency } from '../manifest/dependency';
import { Registration } from './registration';
import * as registry from './registry-manager';
import * as path from './path-manager';
import * as git from './git-manager';

export { Registration, registry, path, git };
export { getRegistrationId } from './registration';

export default class Manager {
  config: Config;

  constructor(config: Config) {
    this.config = config;
  }

  async update() {
    await registry.update(this.config);
  }

  async resolve(dependency: Dependency): Promise<Registration[]> {
    if (isRegistryDependency(dependency)) {
      return registry.resolve(this.config, dependency);
    } else if (isGitDependency(dependency)) {
      return git.resolve(this.config, dependency);
    } else {
      return path.resolve(this.config, dependency);
    }
  }

  async fetch(registration: Registration): Promise<string> {
    const [type] = registration.source.split('+', 1);

    if (type === 'registry') {
      return registry.fetch(this.config, registration);
    } else if (type === 'git') {
      return git.fetch(this.config, registration);
    } else {
      return path.fetch(this.config, registration);
    }
  }
}
