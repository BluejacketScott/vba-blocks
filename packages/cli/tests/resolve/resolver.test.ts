import { loadConfig } from '../../src/config';
import Resolver from '../../src/resolve/resolver';

jest.mock('../../src/sources/registry-source');

const dependency = {
  name: 'a',
  version: '^1.0.0',
  defaultFeatures: true,
  features: []
};

test('should get dependency from clean state', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should get dependency with existing', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  resolver.graph.set('a', { name: 'a', range: [], registered: [] });

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should include preferred with resolved', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);

  resolver.prefer([
    {
      name: 'a',
      id: 'a@1.0.0',
      source: 'registry+<url>#<none>',
      version: '1.0.0',
      features: [],
      dependencies: [
        {
          name: 'd',
          version: '^1.0.0',
          defaultFeatures: true,
          features: []
        }
      ]
    }
  ]);

  const resolved = await resolver.get(dependency);
  expect(resolved).toMatchSnapshot();
});

test('should get registration', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);
  const resolved = await resolver.get(dependency);

  const registration = resolver.getRegistration('a@1.0.0');
  expect(registration).toMatchSnapshot();
});

test('should iterate through graph entries', async () => {
  const config = await loadConfig();
  const resolver = new Resolver(config);
  const resolved = await resolver.get(dependency);

  const graph = [...resolver];
  expect(graph).toMatchSnapshot();
});
