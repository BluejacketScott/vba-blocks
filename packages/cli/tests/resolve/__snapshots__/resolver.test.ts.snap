// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get dependency from clean state 1`] = `
Object {
  "name": "a",
  "range": Array [
    "^1.0.0",
  ],
  "registered": Array [
    Object {
      "dependencies": Array [],
      "features": Array [],
      "id": "a@0.1.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "0.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^1.0.0",
        },
      ],
      "features": Array [],
      "id": "a@1.0.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.0.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^1.2.0",
        },
      ],
      "features": Array [],
      "id": "a@1.1.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^2",
        },
      ],
      "features": Array [],
      "id": "a@1.2.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.2.0",
    },
  ],
}
`;

exports[`should get dependency with existing 1`] = `
Object {
  "name": "a",
  "range": Array [
    "^1.0.0",
  ],
  "registered": Array [],
}
`;

exports[`should get registration 1`] = `
Object {
  "dependencies": Array [
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "name": "d",
      "version": "^1.0.0",
    },
  ],
  "features": Array [],
  "id": "a@1.0.0",
  "name": "a",
  "source": "registry+<url>#<none>",
  "version": "1.0.0",
}
`;

exports[`should include preferred with resolved 1`] = `
Object {
  "name": "a",
  "preferred": Object {
    "dependencies": Array [
      Object {
        "defaultFeatures": true,
        "features": Array [],
        "name": "d",
        "version": "^1.0.0",
      },
    ],
    "features": Array [],
    "id": "a@1.0.0",
    "name": "a",
    "source": "registry+<url>#<none>",
    "version": "1.0.0",
  },
  "range": Array [
    "^1.0.0",
  ],
  "registered": Array [
    Object {
      "dependencies": Array [],
      "features": Array [],
      "id": "a@0.1.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "0.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^1.0.0",
        },
      ],
      "features": Array [],
      "id": "a@1.0.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.0.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^1.2.0",
        },
      ],
      "features": Array [],
      "id": "a@1.1.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.1.0",
    },
    Object {
      "dependencies": Array [
        Object {
          "defaultFeatures": true,
          "features": Array [],
          "name": "d",
          "version": "^2",
        },
      ],
      "features": Array [],
      "id": "a@1.2.0",
      "name": "a",
      "source": "registry+<url>#<none>",
      "version": "1.2.0",
    },
  ],
}
`;

exports[`should iterate through graph entries 1`] = `
Array [
  Array [
    "a",
    Object {
      "name": "a",
      "range": Array [
        "^1.0.0",
      ],
      "registered": Array [
        Object {
          "dependencies": Array [],
          "features": Array [],
          "id": "a@0.1.0",
          "name": "a",
          "source": "registry+<url>#<none>",
          "version": "0.1.0",
        },
        Object {
          "dependencies": Array [
            Object {
              "defaultFeatures": true,
              "features": Array [],
              "name": "d",
              "version": "^1.0.0",
            },
          ],
          "features": Array [],
          "id": "a@1.0.0",
          "name": "a",
          "source": "registry+<url>#<none>",
          "version": "1.0.0",
        },
        Object {
          "dependencies": Array [
            Object {
              "defaultFeatures": true,
              "features": Array [],
              "name": "d",
              "version": "^1.2.0",
            },
          ],
          "features": Array [],
          "id": "a@1.1.0",
          "name": "a",
          "source": "registry+<url>#<none>",
          "version": "1.1.0",
        },
        Object {
          "dependencies": Array [
            Object {
              "defaultFeatures": true,
              "features": Array [],
              "name": "d",
              "version": "^2",
            },
          ],
          "features": Array [],
          "id": "a@1.2.0",
          "name": "a",
          "source": "registry+<url>#<none>",
          "version": "1.2.0",
        },
      ],
    },
  ],
]
`;
